lang/src/org/partiql/lang/eval/physical/operators/AggregateOperatorFactory.kt

```diff
@@ -0,0 +1,172 @@
+import java.util.TreeSet
+
+public abstract class AggregateOperatorFactory(name: String) : RelationalOperatorFactory {
+    public final override val key: RelationalOperatorFactoryKey = RelationalOperatorFactoryKey(RelationalOperatorKind.AGGREGATE, name)
+    public abstract fun create(
+        impl: PartiqlPhysical.Impl,
+        source: RelationExpression,
+        keys: List<CompiledGroupKey>,
+        functions: List<CompiledAggregateFunction>
+    ): RelationExpression
+
+    public abstract fun getFunctionFactory(
+        operator: AggregationKind,
+        name: String
+    ): AggregationFunctionFactory
+}
+
+internal class FullAggregateFactory(name: String) : AggregateOperatorFactory(name) {
+
+    private val aggregationFunctionFactories = mapOf(
+        AggregationFunctionFactoryKey(AggregationKind.MAX, "MAX") to MaxFactory("MAX"),
+        AggregationFunctionFactoryKey(AggregationKind.MIN, "MIN") to MinFactory("MIN"),
+        AggregationFunctionFactoryKey(AggregationKind.COUNT, "COUNT") to CountFactory("COUNT"),
+        AggregationFunctionFactoryKey(AggregationKind.SUM, "SUM") to SumFactory("SUM"),
+        AggregationFunctionFactoryKey(AggregationKind.GROUP_AS, "GROUP_AS") to GroupAsFactory("GROUP_AS"),
+        AggregationFunctionFactoryKey(AggregationKind.AVG, "AVG") to AverageFactory("AVG"),
+    )
+
+    override fun getFunctionFactory(operator: AggregationKind, name: String): AggregationFunctionFactory {
+        val key = AggregationFunctionFactoryKey(operator, name)
+        return aggregationFunctionFactories[key] ?: error("Factory for operator ${key.kind} named '${key.name}' does not exist.")
+    }
+
+    override fun create(
+        impl: PartiqlPhysical.Impl,
+        source: RelationExpression,
+        keys: List<CompiledGroupKey>,
+        functions: List<CompiledAggregateFunction>
+    ): RelationExpression =
+        RelationExpression { state ->
+            val sourceIter = source.evaluate(state)
+
+            val aggregationMap = AggregationMap()
+            val seen = List(functions.size) { TreeSet(DEFAULT_COMPARATOR) }
+            val registerSize = state.registers.size
+            val variables = functions.flatMap { function -> function.variables } + keys.map { key -> key.variable }
+            val variablesIndexes = variables.map { variable -> variable.index.value.toIntExact() }.toSet()
+
+            while (sourceIter.nextRow()) {
+
+                // Initialize the AggregationMap
+                val isMissingArray = Array(keys.size) { false }
+                val evaluatedGroupByKeys = keys.mapIndexed { index, groupByKey ->
+                    val ionValue = groupByKey.value.invoke(state).ionValue
+                    if (ionValue.isNullValue) {
+                        isMissingArray[index] = ionValue.hasTypeAnnotation(MISSING_ANNOTATION)
+                        ionValue.removeTypeAnnotation(MISSING_ANNOTATION)
+                    }
+                    ionValue
```

Comment:
Could you please explain a little more on why do we need to use ionValue here? (as opposite to a `sexp ExprValue`)

Reply:
IonValue implements equals and hashcode, so we can use them as keys within a map.